### PR TITLE
Fix problem with cursor jumping back to start when adding an emoji in edit mode

### PR DESCRIPTION
**Pull request by MrRefactor**
### Details

In the chrome browser on Android devices nad Android app when the user edits the message and chooses an emoji from the emoji modal, the cursor jumps back at the beginning of the input. The problem is that addEmojiToTextbox is called from EmojiPickerButton component's onEmojiSelected callback prop, which is called just before onModalHide. The former in turn calls [focusComposerWithDelay](https://github.com/Expensify/App/blob/8d24b8d737071d7014c622687a3785f149750a9f/src/libs/focusComposerWithDelay.ts#L9), which calls focus() on the text input. That method is annotated with the commend 'The exact behavior triggered will depend on the platform and type of view.', which apparently on Android mWeb Chrome and Android app moves the cursor to the start and is the root cause of the problem in this issue. Adding an additional parameter to this function to force-set the position to the state-carried one on this platform fixes the problem and even - when applied - renders the original proposal solution unnecessary.

### Fixed Issues

$ https://github.com/Expensify/App/issues/28563
PROPOSAL: https://github.com/Expensify/App/issues/28563#issuecomment-1814500908

### Tests

- Send a message to a chat
- Long-press it and select "Edit comment" in the context menu
- Append a new emoji from the emoji picker sliding menu - repeat this in several edge cases: in-between existing text, after text, before text
- Ensure the cursor position is each time in the expected position and does not jump to the start of the text

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as Tests

### QA Steps

Same as Tests

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/32591413/29d59c7f-cdaa-466f-9e6e-f4496444adbc

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/32591413/b8f9bbe0-a716-46bc-8be9-acd8b9e8316f

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/32591413/5f722ac1-88be-4ff3-b6e6-4f01a5d080f2

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/32591413/7d18fc66-e797-4203-8ca9-4f4265596498

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/32591413/c77ecf5b-a29f-4cc5-b38a-20dfa6e42d83

https://github.com/Expensify/App/assets/32591413/9db1c1ba-9776-4207-a0dc-ac8a660d1a1d

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/32591413/7d1b07b1-4bcd-4ec4-a428-c45f5e2dcc83

</details>

### CODE REVIEW

File: src/libs/focusComposerWithDelay/index.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+                return;
+            }
+            textInput.focus();
+            if (forceSetSelection) {
+                if (setSelectionRange) {
+                    (textInput as HTMLTextAreaElement).setSelectionRange(forceSetSelection.start, forceSetSelection.end);
+                } else {
+                    (textInput as TextInput).setSelection(forceSetSelection.start, forceSetSelection.end);
+                }
+            }
```
Comment (TMisiukiewicz) [inline src/libs/focusComposerWithDelay/index.ts:47]: can we move it to a separate function to avoid doubling the same code here and :25-33?

**Review by Swor71 — APPROVED** [review]
lgtm

**Comment by MrRefactor** [comment]
@abdulrahuman5196 @neil-marcellini updated/refactored PR

**Comment by abdulrahuman5196** [comment]
@neil-marcellini I still think this has become a complicated fix for a minor issue. Should we do this way?

File: src/pages/home/report/ReportActionItemMessageEdit.tsx [hunk 2]
```diff
@@ -329,14 +329,22 @@ function ReportActionItemMessageEdit(
         deleteDraft();
     }, [action, debouncedSaveDraft, deleteDraft, draft, reportID]);
 
+    const shouldUseEmojiPickerSelectionRef = shouldUseEmojiPickerSelection();
```
Comment (neil-marcellini) [inline src/pages/home/report/ReportActionItemMessageEdit.tsx:332]: Kind of weird to call this a ref, and I don't think we need to create a variable for this, let's just call it directly.

File: src/pages/home/report/ReportActionItemMessageEdit.tsx [hunk 3]
```diff
@@ -329,14 +329,22 @@ function ReportActionItemMessageEdit(
...
-            end: prevSelection.start + emoji.length + CONST.SPACE_LENGTH,
-        }));
+        const newSelection = {
+            start: selection.start + emoji.length + CONST.SPACE_LENGTH,
+            end: selection.start + emoji.length + CONST.SPACE_LENGTH,
+        };
+        setSelection(newSelection);
+
+        if (shouldUseEmojiPickerSelectionRef) {
+            // immediately set the selection again on android and Chrome mobile after focusing the
```
Comment (neil-marcellini) [inline src/pages/home/report/ReportActionItemMessageEdit.tsx:344]: I find this comment a bit confusing. How about something like this instead?

```ts
// On Android and Chrome mobile, focusing the input sets the cursor position back to the start.
// To fix this, immediately set the selection again after focusing the input.
```

Comment (neil-marcellini) [reply src/pages/home/report/ReportActionItemMessageEdit.tsx:344]: Is that what you are trying to explain?

File: src/pages/home/report/ReportActionItemMessageEdit.tsx [hunk 4]
```diff
@@ -450,7 +458,12 @@ function ReportActionItemMessageEdit(
                     <View style={styles.editChatItemEmojiWrapper}>
                         <EmojiPickerButton
                             isDisabled={shouldDisableEmojiPicker}
-                            onModalHide={() => focus(true)}
+                            onModalHide={() => {
+                                const emojiPickerSelection = emojiPickerSelectionRef.current ? {...emojiPickerSelectionRef.current} : undefined;
+                                emojiPickerSelectionRef.current = undefined;
+
+                                focus(true, emojiPickerSelection);
```
Comment (neil-marcellini) [inline src/pages/home/report/ReportActionItemMessageEdit.tsx:465]: Why do we need to copy the ref and then set it to undefined? Can you please explain with a comment?

**Review by neil-marcellini — CHANGES_REQUESTED** [review]
Leaving a partial review, I'll review more later.

**Review by neil-marcellini — CHANGES_REQUESTED** [review]
Although a lot of files are changed here and I do have some questions and suggestions, overall the approach seems fine and it looks like it creates the proper result.

@abdulrahuman5196 I think we should proceed with the solution, unless you feel that there are major flaws I'm missing.

**Comment by neil-marcellini** [comment]
❗❗ Heads up, I'm going to be OOO working from Spain 🇪🇸 part time until 3/28. Most days I will be working 50%, some days 100%. Please DM me if something needs urgent attention.❗❗

**Comment by MrRefactor** [comment]
Updated PR with changes from review, please continue with the review @neil-marcellini @abdulrahuman5196

**Comment by abdulrahuman5196** [comment]
Reviewing again

**Comment by abdulrahuman5196** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>





https://github.com/Expensify/App/assets/46707890/0341b9a8-2946-4ac8-8e84-580d363d5c33





</details>

<details>
<summary>Android: mWeb Chrome</summary>




https://github.com/Expensify/App/assets/46707890/507cc014-ddb2-47d5-82a4-1cb31385bdd0




</details>

<details>
<summary>iOS: Native</summary>





https://github.com/Expensify/App/assets/46707890/098ed294-8ecc-4020-a868-f7e479018a56





</details>

<details>
<summary>iOS: mWeb Safari</summary>




https://github.com/Expensify/App/assets/46707890/621d06b1-08cf-4543-8724-d31a9b4bbe7e



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>





https://github.com/Expensify/App/assets/46707890/015a3955-4e97-4cba-9413-86c5cd248655


</details>

<details>
<summary>MacOS: Desktop</summary>




https://github.com/Expensify/App/assets/46707890/ae3883da-0a95-456b-81b9-522d6da534d7




</details>

**Review by abdulrahuman5196 — APPROVED** [review]
Changes looks good and works well. Reviewers checklist is also complete.

All yours. @neil-marcellini 

🎀 👀 🎀
C+ Reviewed

**Review by neil-marcellini — APPROVED** [review]
The code looks good to me and it seems to test well. Thanks!

**Pull request merged**